### PR TITLE
DRILL-7333: Set container record count where missing

### PR DESCRIPTION
**Pull request by paul-rogers**
Many operators fail to set the container record count. See DRILL-7333
for details. This commit fixes a large number of operators. Replaces
blocks of copy/paste code to set the container vector value count with
a call to a single function that does the work. That function also
sets the count for the container as a whole, something that many
operators failed to do, causing hard-to-find errors downstream.

Tests: reran all unit tests. A new unit test, to verify that the
count is set, will come in a later commit once all fixes are made.

Also contains code cleanup for items noticed while finding the container count issues.

### CODE REVIEW

**Comment by paul-rogers** [comment]
Several tests fail due a race condition in Drillbit shutdown:

```
[ERROR]  TestUnionAll.testUnionAllEmptySides:348->BaseTestQuery.testRunAndReturn:340 » Rpc
[ERROR] TestUnionDistinct.testUnionDistinctEmptySides:347->BaseTestQuery.testRunAndReturn:340 » Rpc
[ERROR]  TestFastComplexSchema.test5:96->BaseTestQuery.testRunAndReturn:340 » Rpc org.a...
[ERROR] TestFlatten.testFlattenAfterJoin1:526->BaseTestQuery.testRunAndReturn:340 » Rpc
[ERROR] TestFlatten.testFlattenAfterJoin2:537->BaseTestQuery.testRunAndReturn:340 » Rpc
[ERROR]  TestFlatten.testFlattenAfterSort:514->BaseTestQuery.testRunAndReturn:340 » Rpc
```

A typical error stack:

```
[INFO] Running org.apache.drill.exec.store.parquet.ParquetInternalsTest
16:45:18.456 [BitClient-1] ERROR o.a.d.exec.rpc.RpcExceptionHandler - Exception in RPC communication.  Connection: /172.16.10.3:35130 <--> paul-linux/172.16.10.3:31075 (data client).  Closing connection.
java.lang.IllegalStateException: Attempting operation on allocator when allocator is closed.
Allocator(rpc:bit-data) 0/0/8256/9223372036854775807 (res/actual/peak/limit)
  child allocators: 0
  ledgers: 0
  reservations: 0

        at org.apache.drill.exec.memory.BaseAllocator.assertOpen(BaseAllocator.java:130) ~[drill-memory-base-1.17.0-SNAPSHOT.jar:1.17.0-SNAPSHOT]
        at org.apache.drill.exec.memory.BaseAllocator.buffer(BaseAllocator.java:209) ~[drill-memory-base-1.17.0-SNAPSHOT.jar:1.17.0-SNAPSHOT]
        at org.apache.drill.exec.memory.DrillByteBufAllocator.buffer(DrillByteBufAllocator.java:49) ~[drill-memory-base-1.17.0-SNAPSHOT.jar:1.17.0-SNAPSHOT]
        at org.apache.drill.exec.memory.DrillByteBufAllocator.ioBuffer(DrillByteBufAllocator.java:64) ~[drill-memory-base-1.17.0-SNAPSHOT.jar:1.17.0-SNAPSHOT]
        at io.netty.channel.AdaptiveRecvByteBufAllocator$HandleImpl.allocate(AdaptiveRecvByteBufAllocator.java:104) ~[netty-transport-4.0.48.Final.jar:4.0.48.Final]
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:117) ~[netty-transport-4.0.48.Final.jar:4.0.48.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:645) [netty-transport-4.0.48.Final.jar:4.0.48.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:580) [netty-transport-4.0.48.Final.jar:4.0.48.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:497) [netty-transport-4.0.48.Final.jar:4.0.48.Final]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:459) [netty-transport-4.0.48.Final.jar:4.0.48.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:131) [netty-common-4.0.48.Final.jar:4.0.48.Final]
        at java.lang.Thread.run(Thread.java:748) [na:1.8.0_212]
```

This error is similar to the one that causes a UDF test to randomly fail. I'll try to track it down, but since the error is pre-existing, I'll not hold up this PR for that fix.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/StreamingAggBatch.java [hunk 1]
```diff
@@ -75,12 +75,12 @@
 import static org.apache.drill.exec.record.RecordBatch.IterOutcome.STOP;
 
 public class StreamingAggBatch extends AbstractRecordBatch<StreamingAggregate> {
-  static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(StreamingAggBatch.class);
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(StreamingAggBatch.class);
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/StreamingAggBatch.java]: Please import `org.slf4j.Logger` and `org.slf4j.LoggerFactory`.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/filter/FilterRecordBatch.java [hunk 2]
```diff
@@ -164,15 +167,14 @@ protected Filterer generateSV4Filterer() throws SchemaChangeException {
...
-
   }
 
   protected Filterer generateSV2Filterer() throws SchemaChangeException {
     final ErrorCollector collector = new ErrorCollectorImpl();
     final List<TransferPair> transfers = Lists.newArrayList();
     final ClassGenerator<Filterer> cg = CodeGenerator.getRoot(Filterer.TEMPLATE_DEFINITION2, context.getOptions());
-    // Uncomment below lines to enable saving generated code file for debugging
-    // cg.getCodeGenerator().plainJavaCapable(true);
+    cg.getCodeGenerator().plainJavaCapable(true);
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/filter/FilterRecordBatch.java]: Is this change needed? Maybe it was left by mistake...

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/filter/RuntimeFilterRecordBatch.java [hunk 3]
```diff
@@ -53,27 +53,27 @@
  * the Query's performance ,but just do a memory transfer by the later RemovingRecordBatch op.
  */
 public class RuntimeFilterRecordBatch extends AbstractSingleRecordBatch<RuntimeFilterPOP> {
-  private SelectionVector2 sv2;
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(RuntimeFilterRecordBatch.class);
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/filter/RuntimeFilterRecordBatch.java]: Imports...

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java [hunk 4]
```diff
@@ -1048,6 +1059,8 @@ public IterOutcome executeBuildPhase() throws SchemaChangeException {
           read_right_HV_vector = null;
         }
         break;
+      default:
+        throw new IllegalStateException(rightUpstream.name());
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java]: Maybe add more descriptive error message.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java [hunk 5]
```diff
@@ -1250,8 +1263,10 @@ public HashJoinBatch(HashJoinPOP popConfig, FragmentContext context,
     partitions = new HashPartition[0];
 
     // get the output batch size from config.
-    final int configuredBatchSize = (int) context.getOptions().getOption(ExecConstants.OUTPUT_BATCH_SIZE_VALIDATOR);
-    final double avail_mem_factor = (double) context.getOptions().getOption(ExecConstants.OUTPUT_BATCH_SIZE_AVAIL_MEM_FACTOR_VALIDATOR);
+    final int configuredBatchSize = (int) context.getOptions().getOption(
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java]: If option is long why we use int?

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectorTemplate.java [hunk 6]
```diff
@@ -97,6 +97,10 @@ public final void setup(FragmentContext context, RecordBatch incoming, RecordBat
     case TWO_BYTE:
       this.vector2 = incoming.getSelectionVector2();
       break;
+    case NONE:
+      break;
+    default:
+      throw new UnsupportedOperationException();
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectorTemplate.java]: Please add error message.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/unorderedreceiver/UnorderedReceiverBatch.java [hunk 7]
```diff
@@ -230,7 +230,9 @@ public void close() {
 
   @Override
   public VectorContainer getOutgoingContainer() {
-    throw new UnsupportedOperationException(String.format(" You should not call getOutgoingContainer() for class %s", this.getClass().getCanonicalName()));
+    throw new UnsupportedOperationException(
+        String.format("You should not call getOutgoingContainer() for class %s",
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/unorderedreceiver/UnorderedReceiverBatch.java]: I think referring as `you` is not common standard for error messages, please re-write.

File: exec/java-exec/src/main/java/org/apache/drill/exec/record/AbstractUnaryRecordBatch.java [hunk 8]
```diff
@@ -34,9 +34,9 @@
  * @param <T>
  */
 public abstract class AbstractUnaryRecordBatch<T extends PhysicalOperator> extends AbstractRecordBatch<T> {
-  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(new Object() {}.getClass().getEnclosingClass());
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(AbstractUnaryRecordBatch.class);
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/record/AbstractUnaryRecordBatch.java]: Imports.

**Comment by arina-ielchiieva** [comment]
@paul-rogers 
1. travis job fail because there are some unused imports, could you please them and address code review comment so I can run full test suit for you branch.
2. 
> This commit fixes a large number of operators.

Are operators that are not fixed?

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/filter/FilterRecordBatch.java (hunk 2 above)
Comment (paul-rogers) [reply exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/filter/FilterRecordBatch.java]: The rule is supposed to be that all operators that support "plain Java" declare that they can do so. This allows us, at some point, to flip a switch and use the Java compiler rather than Janino and byte code fixup. So, the "plainJavaCapable" is supposed to be uncommented. Only the "saveCodeForDebugging" should be commented. So, this was fixing an earlier error.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/filter/RuntimeFilterRecordBatch.java (hunk 3 above)
Comment (paul-rogers) [reply exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/filter/RuntimeFilterRecordBatch.java]: I wonder, rather than fixing these each time I mark a logger as private, should someone run a script to fix up all the files?

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java (hunk 5 above)
Comment (paul-rogers) [reply exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java]: This is original code, I just split the line for legibility. One could argue that a patch size larger than 2 GB would be excessive, I suppose.

**Comment by paul-rogers** [comment]
@arina-ielchiieva, thank your for your review on a weekend! Addressed the minor comments.

This PR tackles only missing container and record count values.

These errors were found by enabling the "batch validator" and adding many rigorous consistency checks. The validator also checks vectors: there are hundreds of errors in which counts are wrong, offset vectors are invalid and so on. Those will be done as a separate PR to avoid overwhelming the reviewer (or myself.) This is also why the batch validator itself is not part of this PR: it would complain about the invalid vectors.

In starting on the vector fixes, many of these same operators are fixed again, along with a few others.

An earlier comment mentioned that I saw a number of RPC errors due to what seems like a race condition in closing allocators. Oddly, it turns out that Travis did not see those errors, but instead found an out-of-bounds error.

When tests run under Maven, we turn on checking of vector bounds. I discovered that bounds checks are turned off in the IDE, which is why I didn't catch the issue. I'll follow up with another commit for that fix. (That commit will also turn on bounds checking if assertions are enabled so that bounds checking occurs in IDE test runs.)

**Comment by paul-rogers** [comment]
Addressed review comments. Still getting failures when running tests in Maven, but not Eclipse. I guess we should hold off committing this until I track down the issues:

```
[ERROR]   TestUnionAll.testUnionAllEmptySides:348->BaseTestQuery.testRunAndReturn:340 » Rpc
[ERROR]   TestUnionDistinct.testUnionDistinctEmptySides:347->BaseTestQuery.testRunAndReturn:340 » Rpc
[ERROR]   TestFastComplexSchema.test5:96->BaseTestQuery.testRunAndReturn:340 » Rpc org.a...
[ERROR]   TestFlatten.testFlattenAfterJoin1:526->BaseTestQuery.testRunAndReturn:340 » Rpc
[ERROR]   TestFlatten.testFlattenAfterJoin2:537->BaseTestQuery.testRunAndReturn:340 » Rpc
[ERROR]   TestFlatten.testFlattenAfterSort:514->BaseTestQuery.testRunAndReturn:340 » Rpc
[ERROR]   TestDynamicUDFSupport.testReRegisterTheSameJarWithDifferentContent:600->BaseTestQuery.testRunAndReturn:340 » Rpc
```

**Comment by arina-ielchiieva** [comment]
@paul-rogers mostly rpc errors are caused by other tests failures.
Currently if we look in travis there are still IOB exceptions (https://travis-ci.org/apache/drill/jobs/564781376).

**Comment by paul-rogers** [comment]
@arina-ielchiieva, thanks much for the reviews and the suggestions. I would tend to doubt that the RPC errors I'm seeing are due to either lack of resources or other test failures. For example, the UDF tests fail about half the time when I do builds, even clean builds on master. Maven controls resources, the Dynamic UDF tests are the only ones that fail, and my machine has 32 GB of memory; far more than needed. Something odd is going on; I just don't know what.

In any event, I'll take another crack at tracking down the issue. If that does not work, I'll back out the Flatten and selected other changes until I get a clean build. I'll add Flatten to a later commit once I figure out what's what. Will try to complete this work by next Monday.

**Comment by paul-rogers** [comment]
Going to abandon this. Turns out that Drill has many, many interlocking vector count and size errors. I've not found a good way to work on them in chunks, and the problem is too large to do in one go. (I now understand what Jinfeng was facing a couple of years ago in his "empty batches" project!) Since I'm running out of time to work on this, we'll just leave well enough alone.

Started down this route because the revised JSON reader hit a number of these problems as it used the "right" values which broke code that expected the wrong values. Since there is no time to work on fixing this, it means that the JSON reader can't be offered. Will revisit later if time becomes available.
